### PR TITLE
Added the ability to add a custom watermark

### PR DESCRIPTION
**Pull request by kaicianflone**
Added the ability to upload a custom watermark from issue #395 

Currently only square watermarks will work and I have them scaled down to 75px * 75px.

Currently your watermark will also save as the watermark setting until you press reset default settings too 👍

### CODE REVIEW

File: components/Settings.js [hunk 1]
```diff
@@ -105,6 +106,7 @@ class Settings extends React.Component {
               enabled={this.props.watermark}
               onChange={this.props.onChange.bind(null, 'watermark')}
             />
+            {this.props.watermark && <CustomWatermarkPicker onChange={this.props.onUpdate} />}
```
Comment (mfix22) [inline components/Settings.js]: Just use `this.props. onChange` like the other settings do.

File: components/Carbon.js [hunk 2]
```diff
@@ -106,6 +106,18 @@ class Carbon extends PureComponent {
...
       </div>
     )
+    let watermark = <Watermark />
+    if (this.props.customWatermark) {
+      let watermarkSource = this.props.customWatermark
+      watermark = (
+        <img
+          className="watermark"
+          style={{ height: '75px', width: '75px' }}
+          src={watermarkSource.customWatermark}
```
Comment (mfix22) [inline components/Carbon.js]: You have `this.props.customWatermark.customWatermark.` . . . are you able to flatten this? (:

File: components/CustomWatermarkPicker.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  constructor(props) {
+    super(props)
+    this.selectImage = this.selectImage.bind(this)
+  }
+
+  selectImage(e) {
+    const file = e.target ? e.target.files[0] : e
+
+    return fileToDataURL(file).then(dataURL =>
+      this.setState(() => {
```
Comment (mfix22) [inline components/CustomWatermarkPicker.js]: Why are you calling `setState` here? You don't need to if you don't have any state.

File: lib/constants.js [hunk 4]
```diff
@@ -487,5 +487,6 @@ export const DEFAULT_SETTINGS = {
   exportSize: '2x',
   titleBar: '',
   watermark: false,
+  customWatermark: false,
```
Comment (mfix22) [inline lib/constants.js]: I would make this `null` since it doesn't become a boolean property.

File: components/CustomWatermarkPicker.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    return (
+      <div className={`toggle`}>
+        <span>
+          <input type="file" onChange={this.selectImage} />{' '}
+          <span className="labelText"> Upload Custom Watermark </span>
+        </span>
+
+        <style jsx>
+          {`
+            .toggle {
```
Comment (mfix22) [inline components/CustomWatermarkPicker.js]: @jakedex  mind reviewing this for style changes whenever you get a sec?

**Comment by kaicianflone** [comment]
Ok thank you for the review @mfix22 :) That was my first PR and a very good learning experience. I am going to add some testing too. Thanks!

File: components/CustomWatermarkPicker.js [hunk 6]
```diff
@@ -10,20 +10,18 @@ export default class extends React.Component {
...
   }
 
   render() {
     return (
       <div className={`toggle`}>
         <span>
-          <input type="file" onChange={this.selectImage} />{' '}
+          <input
+            type="file"
+            accept="image/x-png,image/jpeg,image/jpg,image/svg+xml"
```
Comment (mfix22) [inline components/CustomWatermarkPicker.js]: `image/*` (:

File: components/CustomWatermarkPicker.js [hunk 7]
```diff
@@ -10,20 +10,18 @@ export default class extends React.Component {
...
     const file = e.target ? e.target.files[0] : e
 
-    return fileToDataURL(file).then(dataURL =>
-      this.setState(() => {
-        this.props.onChange({
-          customWatermark: dataURL
-        })
-      })
-    )
+    return fileToDataURL(file).then(customWatermark => this.props.onChange(customWatermark))
```
Comment (mfix22) [inline components/CustomWatermarkPicker.js]: ` fileToDataURL(file).then(this.props.onChange)`

**Comment by mfix22** [comment]
@kaicianflone this is looking a lot better! These merge conflicts ^^ should be super easy to fix (sorry about that). Also what do you think about just making the watermark itself the input you click on to change the image? That way we don't need to crowd the settings and it is more natural to change the image?

**Comment by kaicianflone** [comment]
Oh yeah I like that idea. It should be very smooth. I will go ahead and make the change :)

Also I wanted to ask if I should style the Height/Width to 75 x 75 px or if I should just set the maxHeight / maxWidth to 75px or 100px perhaps if anyone wanted to do a 4:3 watermark instead of the 1:1. What do you think?

**Comment by mfix22** [comment]
@kaicianflone  i think you should do a max-width and max-height yeah 👍

**Comment by mfix22** [comment]
hey @kaicianflone 🙂where are you at with this PR?

**Comment by mfix22** [comment]
Hey @kaicianflone we are really excited about this feature. Are you able to update it based on what we discussed above? 🙂

**Comment by mfix22** [comment]
Closing as stale. Please reopen if you have changes 👍

**Pull request closed without merging**